mkosi/config.py

```diff
@@ -0,0 +1,264 @@
+
+        for section in parser.sections():
+            for k, v in parser.items(section):
+                if not (s := self.lookup.get(k)):
+                    die(f"Unknown setting {k}")
+
+                s.parse(s.dest, v, namespace)
+
+        if extras and path.parent.joinpath("mkosi.conf.d").exists():
+            for p in sorted(path.parent.joinpath("mkosi.conf.d").iterdir()):
+                if p.is_dir() or p.suffix == ".conf":
+                    self._parse_config(p, namespace)
+
+
+    def parse(self, namespace: argparse.Namespace = argparse.Namespace()) -> argparse.Namespace:
+        self._parse_config(self.directory, namespace)
+        return namespace
+
+
+def config_make_action(settings: Sequence[MkosiConfigSetting]) -> Type[argparse.Action]:
+    lookup = {s.dest: s for s in settings}
+
+    class MkosiAction(argparse.Action):
+        def __call__(
+            self,
+            parser: argparse.ArgumentParser,
+            namespace: argparse.Namespace,
+            values: Union[str, Sequence[Any], None],
+            option_string: Optional[str] = None
+        ) -> None:
+            assert option_string is not None
+
+            if values is None and self.nargs == "?":
+                values = self.const or "yes"
+
+            try:
+                s = lookup[self.dest]
+            except KeyError:
+                die(f"Unknown setting {option_string}")
+
+            if values is None or isinstance(values, str):
+                s.parse(self.dest, values, namespace)
+            else:
+                for v in values:
+                    assert isinstance(v, str)
+                    s.parse(self.dest, v, namespace)
+
+    return MkosiAction
+
+
+def make_enum_parser(type: Type[enum.Enum]) -> Callable[[str], enum.Enum]:
+    def parse_enum(value: str) -> enum.Enum:
+        try:
+            return type[value]
+        except KeyError:
+            die(f"Invalid enum value {value}")
+
+    return parse_enum
+
+
+def config_make_enum_parser(type: Type[enum.Enum]) -> ConfigParseCallback:
+    def config_parse_enum(dest: str, value: Optional[str], namespace: argparse.Namespace) -> None:
+        setattr(namespace, dest, make_enum_parser(type)(value) if value is not None else value)
+
+    return config_parse_enum
+
+
+def config_make_enum_matcher(type: Type[enum.Enum]) -> ConfigMatchCallback:
+    def config_match_enum(dest: str, value: str, namespace: argparse.Namespace) -> bool:
+        return cast(bool, make_enum_parser(type)(value) == getattr(namespace, dest))
+
+    return config_match_enum
+
+
+def config_make_list_parser(delimiter: str, parse: Callable[[str], Any] = str) -> ConfigParseCallback:
+    def config_parse_list(dest: str, value: Optional[str], namespace: argparse.Namespace) -> None:
+        if not value:
+            setattr(namespace, dest, [])
+            return
+
+        value = value.replace(os.linesep, delimiter)
+        values = [v for v in value.split(delimiter) if v]
```

Comment:
While `os.linesep` is the correct thing, we don't support `mkosi` on anything other than running on Linux and can hardcode `\n` to make this easier to follow.

Alternatively, can this could also be
```suggestion
        values = [v for line in value.splitlines() for v in line.split(delimiter) if v]
```